rust-spec/lightclient/detection/detection.md

```diff
@@ -105,80 +139,194 @@ time t there exists an *h* and a sequence *a(1)*, ... *a(h)* s.t.
+next validators) of some previous block deviated from the protocol. The
+upcoming "accountability" specification will describe how to compute
+a set of at least 1/3 faulty nodes from two conflicting blocks. []
+
+There are different ways to characterize forks
+and attack scenarios. This specification uses the "node-based
+characterization of attacks" which focuses on what kinds of nodes are
+affected (light nodes vs. full nodes). For future reference and
+discussion we also provide a
+"block-based characterization of attacks" below.
+
+### Node-based characterization of attacks
+
+#### **[TMBC-MC-FORK.1]**
+
+We say there is a (main chain) fork at time *t* if
+
+- there are two correct full nodes *i* and *j* and
+- *i* is different from *j* and
+- *i* has decided on *b* and
+- *j* has decided on *c* and
+- there exist *a* such that *attack(a,b,c,t)*.
+
+#### **[TMBC-LC-ATTACK.1]**
+
+We say there is a light client attack at time *t*, if
+
+- there is **no** (main chain) fork [[TMBC-MC-FORK.1]](#TMBC-MC-FORK1), and
+- there exist nodes that have computed light blocks *b* and *c* and
+- there exist *a* such that *attack(a,b,c,t)*.
+
+We say the attack is at height *a.Header.Height*.
+
+> In this specification we consider detection of light client
+> attacks. Intuitively, the case we consider is that
+> light block *b* is the one from the
+> blockchain, and some attacker has computed *c* and tries to wrongly
+> convince
+> the light client that *c* is the block from the chain.
+
+#### **[TMBC-LC-ATTACK-EVIDENCE.1]**
+
+We consider the following case of a light client attack
+[[TMBC-LC-ATTACK.1]](#TMBC-LC-ATTACK1):
+
+- *attack(a,b,c,t)*
+- there is a peer p1 that has a sequence *chain* of blocks from *a* to *b*
+- *skip-root(a,c,t)*: by [[TMBC-SKIP-ROOT.1]](#TMBC-SKIP-ROOT1) there is a
+  verification trace *v* of the form *a = v(1)*, ... *v(h) = c*
+
+Evidence for p1 (that proves an attack) consists for index i
+of v(i) and v(i+1) such that
+
+- E1(i). v(i) is equal to the block of *chain* at height v(i).Height, and
+- E2(i). v(i+1) that is different from  the block of *chain* at
+  height v(i+1).height
+
+> Observe p1 can
+>
+> - check that v(i+1)  differs from its block at that height, and
+> - verify v(i+1) in one step from v(i) as v is a verification trace.
+
+**Proposition.** In the case of attack, evidence exists.  
+*Proof.* First observe that
+
+- (A). (NOT E2(i)) implies E1(i+1)
+
+Now by contradiction assume there is no evidence. Thus
+
+- for all i, we have NOT E1(i) or NOT E2(i)
+- for i = 1 we have E1(1) and thus NOT E2(1)
+  thus by induction on i, by (A) we have for all i that **E1(i)**
+- from attack we have E2(h-1), and as there is no evidence for
+  i = h - 1 we get **NOT E1(h-1)**. Contradiction.
+QED.
+
+#### **[TMBC-LC-EVIDENCE-DATA.1]**
+
+To prove the attack to p1, because of Point E1, it is sufficient to
+submit
+
+- v(i).Height (rather than v(i)).
```

Comment:
Callum made an observation that in case of Equivocation and Amnesia conflicting blocks are for the same height and they have the same validator set. So we don't need root height, i.e., we could set root height to v(i+1).Height. But maybe this can be seen as an optimisation (I think this is how it is being implemented in Go version).

Reply:
I don't see what the "optimization" here would be. From a semantic viewpoint, currently the evidence says:  
"Some of the **next validators** in v(i).Height messed up v(i+1)"

If we would change that the semantics it would say:
"Some of the next validators in v(i).Height messed up v(i+1), unless v(i).Height = v(i+1).Height in which case the **validators** in v(i).Height messed up v(i+1)"

My feeling is that it is more natural to treat the special case "v(i).Height + 1= p v(i+1).Height" on the receiving side of the evidence rather on the sending side, and check this condition there.

Also, we do not check at the light client what kind of attack we have  (we decided to not add this information in the evidence). So it is not clear how to figure out that case.

@cmwaters : do I miss something here?